File: languages/java/src/main/java/com/bloock/sdk/client/IntegrityClient.java
package com.bloock.sdk.client;

import com.bloock.sdk.bridge.Bridge;
import com.bloock.sdk.bridge.proto.BloockConfig.ConfigData;
import com.bloock.sdk.bridge.proto.BloockIntegrity.*;
import com.bloock.sdk.bridge.proto.BloockShared.Error;
import com.bloock.sdk.config.Config;
import com.bloock.sdk.entity.integrity.Anchor;
import com.bloock.sdk.entity.integrity.Network;
import com.bloock.sdk.entity.integrity.Proof;
import com.bloock.sdk.entity.integrity.RecordReceipt;
import com.bloock.sdk.entity.record.Record;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Provides functionality to interact with the
 * <a href="https://dashboard.bloock.com/login">Bloock Integrity service</a>.
 */
public class IntegrityClient {
  private final Bridge bridge;
  private final ConfigData configData;

  /**
   * Creates a new IntegrityClient with default configuration.
   */
  public IntegrityClient() {
    this.bridge = new Bridge();
    this.configData = Config.newConfigDataDefault();
  }

  /**
   * Creates a new IntegrityClient with the given configuration.
   * 
   * @param configData
   */
  public IntegrityClient(ConfigData configData) {
    this.bridge = new Bridge();
    this.configData = Config.newConfigData(configData);
  }

  /**
   * Sends records to the Bloock Integrity service for certification.
   * 
   * @param records
   * @return
   * @throws Exception
   */
  public List<RecordReceipt> sendRecords(List<Record> records) throws Exception {
    SendRecordsRequest request = SendRecordsRequest.newBuilder()
        .setConfigData(this.configData)
        .addAllRecords(records.stream().map(x -> x.toProto()).collect(Collectors.toList()))
        .build();

    SendRecordsResponse response = bridge.getIntegrity().sendRecords(request);

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return response.getRecordsList().stream()
        .map(x -> RecordReceipt.fromProto(x))
        .collect(Collectors.toList());
  }

  /**
   * Gets an anchor by its ID from the Bloock Integrity service.
   * 
   * @param id
   * @return
   * @throws Exception
   */
  public Anchor getAnchor(long id) throws Exception {
    GetAnchorRequest request = GetAnchorRequest.newBuilder().setConfigData(this.configData).setAnchorId(id).build();

    GetAnchorResponse response = bridge.getIntegrity().getAnchor(request);

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return Anchor.fromProto(response.getAnchor());
  }

  /**
   * Waits for the completion of an anchor on the Bloock Integrity service.
   * 
   * @param id
   * @return
   * @throws Exception
   */
  public Anchor waitAnchor(long id) throws Exception {
    return waitAnchor(id, 120000);
  }

  /**
   * Waits for the completion of an anchor on the Bloock Integrity service.
   * 
   * @param id
   * @param timeout
   * @return
   * @throws Exception
   */
  public Anchor waitAnchor(long id, long timeout) throws Exception {
    WaitAnchorRequest request = WaitAnchorRequest.newBuilder()
        .setConfigData(this.configData)
        .setAnchorId(id)
        .setTimeout(timeout)
        .build();

    WaitAnchorResponse response = bridge.getIntegrity().waitAnchor(request);

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return Anchor.fromProto(response.getAnchor());
  }

  /**
   * Gets a proof for a set of records from the Bloock Integrity service.
   * 
   * @param records
   * @return
   * @throws Exception
   */
  public Proof getProof(List<Record> records) throws Exception {
    GetProofRequest request = GetProofRequest.newBuilder()
        .setConfigData(this.configData)
        .addAllRecords(records.stream().map(x -> x.toProto()).collect(Collectors.toList()))
        .build();

    GetProofResponse response = bridge.getIntegrity().getProof(request);

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return Proof.fromProto(response.getProof());
  }

  /**
   * Verifies the integrity of a proof.
   * 
   * @param proof
   * @return
   * @throws Exception
   */
  public String verifyProof(Proof proof) throws Exception {
    VerifyProofRequest request = VerifyProofRequest.newBuilder()
        .setConfigData(this.configData)
        .setProof(proof.toProto())
        .build();

    VerifyProofResponse response = bridge.getIntegrity().verifyProof(request);

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return response.getRecord();
  }

  /**
   * Verifies the integrity of a set of records.
   * 
   * @param records
   * @return
   * @throws Exception
   */
  public long verifyRecords(List<Record> records) throws Exception {
    return verifyRecords(records, null);
  }

  /**
   * Verifies the integrity of a set of records for a given network.
   * 
   * @param records
   * @param network
   * @return
   * @throws Exception
   */
  public long verifyRecords(List<Record> records, Network network) throws Exception {
    VerifyRecordsRequest.Builder builder = VerifyRecordsRequest.newBuilder()
        .setConfigData(this.configData)
        .addAllRecords(records.stream().map(x -> x.toProto()).collect(Collectors.toList()));

    if (network != null) {
      builder.setNetwork(network.toProto());
    }

    VerifyRecordsRequest request = builder.build();
    VerifyRecordsResponse response = bridge.getIntegrity().verifyRecords(request);

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return response.getTimestamp();
  }

  /**
   * Validates the integrity of a merkle root proof on blockchain.
   * 
   * @param root
   * @param network
   * @return
   * @throws Exception
   */
  public long validateRoot(String root, Network network) throws Exception {
    ValidateRootRequest request = ValidateRootRequest.newBuilder()
        .setConfigData(this.configData)
        .setRoot(root)
        .setNetwork(network.toProto())
        .build();

    ValidateRootResponse response = bridge.getIntegrity().validateRoot(request);

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return response.getTimestamp();
  }
}

File: languages/php/Bloock/Client/IntegrityClient.php
<?php

namespace Bloock\Client;

use Bloock\Bridge\Bridge;
use Bloock\Config\Config;
use Bloock\ConfigData;
use Bloock\Entity\Integrity\Anchor;
use Bloock\Entity\Integrity\Network;
use Bloock\Entity\Integrity\Proof;
use Bloock\Entity\Integrity\RecordReceipt;
use Bloock\GetAnchorRequest;
use Bloock\GetProofRequest;
use Bloock\SendRecordsRequest;
use Bloock\ValidateRootRequest;
use Bloock\VerifyProofRequest;
use Bloock\VerifyRecordsRequest;
use Bloock\WaitAnchorRequest;
use Exception;

/**
 * Provides functionality to interact with the [Bloock Integrity service](https://dashboard.bloock.com/login).
 */
class IntegrityClient
{
    private $bridge;
    private $config;

    /**
     * Creates a new IntegrityClient with the given configuration.
     * @param ConfigData|null $config
     */
    public function __construct(ConfigData $config = null)
    {
        $this->bridge = new Bridge();
        if ($config != null) {
            $this->config = Config::newConfigData($config);
        } else {
            $this->config = Config::newConfigDataDefault();
        }
    }

    /**
     * Sends records to the Bloock Integrity service for certification.
     * @param array $records
     * @return array
     * @throws Exception
     */
    public function sendRecords(array $records): array
    {
        $req = new SendRecordsRequest();
        $r = [];
        foreach ($records as $record) {
            $r[] = $record->toProto();
        }
        $req->setConfigData($this->config)->setRecords($r);

        $res = $this->bridge->integrity->SendRecords($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }
        $receipts = [];
        foreach ($res->getRecords() as $item) {
            $receipts[] = RecordReceipt::fromProto($item);
        }
        return $receipts;
    }

    /**
     * Gets an anchor by its ID from the Bloock Integrity service.
     * @param int $id
     * @param int $timeout
     * @return Anchor
     * @throws Exception
     */
    public function waitAnchor(int $id, int $timeout = 120000): Anchor
    {
        $req = new WaitAnchorRequest();
        $req->setConfigData($this->config)->setAnchorId($id)->setTimeout($timeout);

        $res = $this->bridge->integrity->WaitAnchor($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return Anchor::fromProto($res->getAnchor());
    }

    /**
     * Waits for the completion of an anchor on the Bloock Integrity service.
     * @param int $id
     * @return Anchor
     * @throws Exception
     */
    public function getAnchor(int $id): Anchor
    {
        $req = new GetAnchorRequest();
        $req->setConfigData($this->config)->setAnchorId($id);

        $res = $this->bridge->integrity->GetAnchor($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return Anchor::fromProto($res);
    }

    /**
     * Gets a proof for a set of records from the Bloock Integrity service.
     * @param array $records
     * @return Proof
     * @throws Exception
     */
    public function getProof(array $records): Proof
    {
        $req = new GetProofRequest();
        $r = [];
        foreach ($records as $record) {
            $r[] = $record->toProto();
        }
        $req->setConfigData($this->config)->setRecords($r);

        $res = $this->bridge->integrity->GetProof($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return Proof::fromProto($res->getProof());
    }

    /**
     * Verifies the integrity of a proof.
     * @param Proof $proof
     * @return string
     * @throws Exception
     */
    public function verifyProof(Proof $proof): string
    {
        $req = new VerifyProofRequest();
        $req->setConfigData($this->config)->setProof($proof->toProto());

        $res = $this->bridge->integrity->VerifyProof($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return $res->getRecord();
    }

    /**
     * Verifies the integrity of a set of records.
     * @param array $records
     * @param string|null $network
     * @return int
     * @throws Exception
     */
    public function verifyRecords(array $records, string $network = null): int
    {
        $req = new VerifyRecordsRequest();
        $r = [];
        foreach ($records as $record) {
            $r[] = $record->toProto();
        }
        $req->setConfigData($this->config)->setRecords($r);

        if ($network != null) {
            $req->setNetwork(Network::toProto($network));
        }

        $res = $this->bridge->integrity->VerifyRecords($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return $res->getTimestamp();
    }

    /**
     * Validates the integrity of a merkle root proof on blockchain.
     * @param string $root
     * @param string $network
     * @return int
     * @throws Exception
     */
    public function validateRoot(string $root, string $network): int
    {
        $req = new ValidateRootRequest();
        $req->setConfigData($this->config)->setRoot($root)->setNetwork(Network::toProto($network));

        $res = $this->bridge->integrity->ValidateRoot($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return $res->getTimestamp();
    }
}